src/analysis/asymptotics.lean

```diff
@@ -0,0 +1,438 @@
+the Fréchet derivative.)
+-/
+import analysis.normed_space.basic
+open filter
+
+namespace asymptotics
+
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+
+section
+variables [has_norm β] [has_norm γ] [has_norm δ]
+
+def is_bigo (f : α → β) (g : α → γ) (l : filter α) : Prop :=
+∃ c, { x | ∥ f x ∥ ≤ c * ∥ g x ∥ } ∈ l.sets
+
+def is_littleo (f : α → β) (g : α → γ) (l : filter α) : Prop :=
+∀ c, { x | c * ∥ f x ∥ ≤ ∥ g x ∥ } ∈ l.sets
+
+theorem is_bigo.comp {f : α → β} {g : α → γ} {l : filter α} (hfg : is_bigo f g l)
+    {δ : Type*} (k : δ → α) :
+  is_bigo (f ∘ k) (g ∘ k) (l.comap k) :=
+let ⟨c, hfgc⟩ := hfg in
+⟨c, mem_comap_sets.mpr ⟨_, hfgc, set.subset.refl _⟩⟩
+
+theorem is_littleo.comp {f : α → β} {g : α → γ} {l : filter α} (hfg : is_littleo f g l)
+    {δ : Type*} (k : δ → α) :
+  is_littleo (f ∘ k) (g ∘ k) (l.comap k) :=
+λ c, mem_comap_sets.mpr ⟨_, hfg c, set.subset.refl _⟩
+
+theorem is_bigo.mono {f : α → β} {g : α → γ} {l₁ l₂ : filter α} (h : l₁ ≤ l₂)
+  (h' : is_bigo f g l₂) : is_bigo f g l₁ :=
+let ⟨c, h'c⟩ := h' in ⟨c, h (h'c)⟩
+
+theorem is_littleo.mono {f : α → β} {g : α → γ} {l₁ l₂ : filter α} (h : l₁ ≤ l₂)
+  (h' : is_littleo f g l₂) : is_littleo f g l₁ :=
+λ c, h (h' c)
+
+theorem is_littleo.to_is_bigo {f : α → β} {g : α → γ} {l : filter α} (hgf : is_littleo f g l) :
+  is_bigo f g l :=
+by { use 1, convert hgf 1, ext x, rw [one_mul, one_mul] }
+
+end
+
+section
+
+variables [has_norm β] [normed_group γ]
+
+theorem is_bigo_zero {g : α → γ} {l : filter α} :
+  is_bigo (λ x, (0 : γ)) g l :=
+⟨0, by { filter_upwards [univ_mem_sets], intros x _, simp }⟩
+
+theorem is_littleo_zero {g : α → γ} {l : filter α} :
+  is_littleo (λ x, (0 : γ)) g l :=
+λ c, by { filter_upwards [univ_mem_sets], intros x _, simp }
+
+@[simp]
+theorem is_bigo_norm_right {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo f (λ x, ∥g x∥) l ↔ is_bigo f g l :=
+by simp only [is_bigo, norm_norm]
+
+@[simp]
+theorem is_littleo_norm_right {f : α → β} {g : α → γ} {l : filter α} :
+  is_littleo f (λ x, ∥g x∥) l ↔ is_littleo f g l :=
+by simp only [is_littleo, norm_norm]
+
+@[simp]
+theorem is_bigo_neg_right {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo f (λ x, -(g x)) l ↔ is_bigo f g l :=
+by { rw ←is_bigo_norm_right, simp only [norm_neg], rw is_bigo_norm_right }
+
+@[simp]
+theorem is_littleo_neg_right {f : α → β} {g : α → γ} {l : filter α} :
+  is_littleo f (λ x, -(g x)) l ↔ is_littleo f g l :=
+by { rw ←is_littleo_norm_right, simp only [norm_neg], rw is_littleo_norm_right }
+
+theorem is_bigo_iff_pos {f : α → β} {g : α → γ} {l : filter α} :
+  is_bigo f g l ↔ ∃ c > 0, { x | ∥f x∥ ≤ c * ∥g x∥ } ∈ l.sets :=
+suffices is_bigo f g l → ∃ c > 0, { x | ∥f x∥ ≤ c * ∥g x∥ } ∈ l.sets,
+  from ⟨this, λ ⟨c, cpos, hc⟩, ⟨c, hc⟩⟩,
+assume ⟨c, hc⟩,
+or.elim (lt_or_ge 0 c)
+  (assume : c > 0, ⟨c, this, hc⟩)
+  (assume h'c : c ≤ 0,
+    have {x : α | ∥f x∥ ≤ 1 * ∥g x∥} ∈ l.sets,
```

Comment:
minor suggestion: I think this proof will be shorter if you use `max c 1` as the existential witness

Reply:
I'll take a look, but I don't think that helps.